service/mongodb/controller.go

```diff
@@ -163,19 +164,54 @@ func (c Controller) GetFilePermissions(ctx context.Context,
 // otherwise returns nil and any error if occurred.
 func (c Controller) GetUserPermissions(
 	ctx context.Context,
-	userID string) ([]*pb.GetUserPermissionsResponse_FileRole, error) {
-	filter := bson.D{
+	userID string, pageNum int64, pageSize int64, isShared bool, appID string) (*pb.GetUserPermissionsResponse, error) {
+
+	var filter bson.D
+
+	filter = append(filter, bson.E{
+		Key:   PermissionBSONUserIDField,
+		Value: userID,
+	})
+
+	if isShared {
+		filter = append(filter, bson.E{
+			Key:   PermissionBSONCreatorField,
+			Value: bson.M{"$ne": userID},
+		})
+	}
+
+	if appID != "" {
+		filter = append(filter, bson.E{
+			Key:   PermissionBSONAppIDField,
+			Value: appID,
+		})
+	}
+
+	// Check if one is negative and the other is not
+	if pageNum < 0 || pageSize < 0 {
+		return nil, fmt.Errorf("pageNum %d and pageSize %d must both be non-negative", pageNum, pageSize)
+	}
+	sort := bson.D{
 		bson.E{
-			Key:   PermissionBSONUserIDField,
-			Value: userID,
+			Key:   MongoObjectIDField,
+			Value: 1,
 		},
 	}
 
-	permissions, err := c.store.GetAll(ctx, filter)
+	// Get permissions by page, sorted by mongoID
+	pageRes, err := c.store.GetUserPermissionsByPage(ctx, pageNum, pageSize, sort, filter)
 	if err != nil {
 		return nil, err
 	}
 
+	// Go over the page of permissions received and reformat them
+	filePermissions := c.reformatFilePermissions(pageRes.permissions)
+
+	return &pb.GetUserPermissionsResponse{Permissions: filePermissions, ItemCount: pageRes.itemCount, PageNum: pageRes.pageNum}, nil
+
+}
+
+func (c Controller) reformatFilePermissions(permissions []service.Permission) []*pb.GetUserPermissionsResponse_FileRole {
 	filePermissions := make([]*pb.GetUserPermissionsResponse_FileRole, 0, len(permissions))
 	for _, permission := range permissions {
```

Comment:
_Performance suggestion improvement:_  You can change the **single-threaded for-loop** to **multi-thread for-loop**. use Go’s built-in goroutines combined with a `sync.WaitGroup` - [Multi-thread For Loops Easily and Safely in Go](https://medium.com/@greenraccoon23/multi-thread-for-loops-easily-and-safely-in-go-a2e915302f8b)

Reply:
After some serious research and bench-marking, It seems that `sync.WaitGroup` is only relevant for scales of running 1000 functions who take between 100 to 1000 nanoseconds each. 
In this function, we are talking about scales of 10 ms tops. 
The async code multiplies the time it takes to handle the array by a factor of 4.
If you are interested in seeing the benchmarks (not prettified), you can watch it here:
https://github.com/meateam/permission-service/compare/feature/pagination...testing/goroutines